cli(node): Add node flags to CLI

PlayMa256 opened the pull request:
**What kind of change does this PR introduce?**
Feature

**Did you add tests for your changes?**
Test added, i'm not really sure if that it is enough. If it is not, let me know. Since node has quite a few flags, i don't think testing each of them is a good thing.

**If relevant, did you update the documentation?**
To be updated. Just double checking: this update has to go in the webpack docs, or internal docs?

**Summary**
This is a feature implementation based on #289. 
This pr introduces a way to pass down node flags to execute the cli. To be able to differenciate these flags, a prefix is added to them. E.g. `--node.max_old_value=2048`. At the end, this allows us to accept every single node flag, without having to compare every single one.

**Does this PR introduce a breaking change?**
no

ematipico commented:
This is awesome! There's a problem with your commit message, that's the reason of the broken build. Try to amend your commit and remove the "." (full stop)

PlayMa256 commented:
Ok! I was not sure why it was breaking though. I'm fixing it.

Enviado a partir do VMware Boxer

Em 07/04/2018 14:26, Emanuele <notifications@github.com> escreveu:

This is awesome! There's a problem with your commit message, that's the reason of the broken build. Try to amend your commit and remove the "." (full stop)

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/webpack/webpack-cli/pull/377#issuecomment-379485725>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AAyvTJJm1CYoaxTeN2tT5lnsBKlQpSblks5tmPbNgaJpZM4TK5nO>.

ematipico commented:
Is your branch rebased from `master`? I don't get why appveyor build is failing

PlayMa256 commented:
Me neither... let me try once again rebasing this, i had already. Just a second.

PlayMa256 commented:
@ematipico now it should be updated. By some reason when i ran `git rebase master` it was not finding differences... anyways. Let's just wait for the cis to run. I'm going to keep my eyes on this.

PlayMa256 commented:
@ematipico it is going to fail again though! As you can [see](http://prntscr.com/j2wyq9) it has been rebased with master.

ev1stensberg commented:
Regarding the rationale around accessing node flags in webpack. This is good for us, cause people use webpack inside Docker, CI's etc. It's pretty nice when you can tap directly into v8 or nodeJS for profiling or perf optimizations. I'm in favor for this.

ev1stensberg commented:
@bmeurer thoughts?

montogeek commented:
@ev1stensberg If the can pass flags to `webpack` bin, can as well pass them to `node`.

ev1stensberg commented:
@montogeek not necessarily. The process won't pick the flags up

montogeek commented:
@ev1stensberg How so?

PlayMa256 commented:
What @montogeek is trying to say is: if someone can call `npm run build` || `npm run something` they can call: `node --max_old_memory=1024....` as well.

boneskull commented:
See https://github.com/nodejs/node/pull/19335

bmeurer commented:
I don't have a strong opinion here, but I tend to agree with @PlayMa256.

boneskull commented:
Mocha does this same thing.  Why?  The ability to pass flags directly to the executable is a better user experience.

ematipico commented:
I really like this feature and as @boneskull  pointed out, the user experience improves

montogeek commented:
@tj What you think about this change? You did it in mocha

PlayMa256 commented:
After further analysis on appveyor, i found it is [running out of memory](https://ci.appveyor.com/project/ev1stensberg/webpack-cli/build/590#L599). I just could not see where it is breaking, since it barely started running tests.

ematipico commented:
@PlayMa256, you could try an experiment. Run tests with current branch and see the amount of memory used. Run tests from master and see the amount memory used. If they differ too much, the problem you be into the code of this PR.

PlayMa256 commented:
@ematipico i've ran the tests and compared the heap consumption for each of them, i've compared master and this branch, both are almost the same (considering the marging of error...)

I removed the tests i added. Let me see how ci behaves.

PlayMa256 commented:
After further investigation, i could not find from where this possible "memory leak" is coming from. I gathered the amount of memory being used by each test, from master and from this branch, both tests are the same. Locally the tests run perfect, without any problem. 

At first i suspected that every single command you ran was going though [this](https://github.com/webpack/webpack-cli/pull/377/files#diff-877a99a662829398e8136ee938cf1c93R37) if statement, but endedup up not being true. Could someone give me a hand on this? I ran out of ideas on how to debug this properly.

ematipico commented on bin/process-node-flags.js:
```diff
@@ -0,0 +1,36 @@
+const spawn = require("cross-spawn");
+const path = require("path");
+
+module.exports = function(argv) {
```
Could you add some jsdoc comments here?

ematipico commented on bin/webpack.js:
```diff
@@ -32,6 +32,13 @@
+
+	if (nodeFlags.length) {
+		require("./process-node-flags")(process.argv);
+		return;
```
Why there's a `return` here?

PlayMa256 replied:
Same reason as used [here](https://github.com/webpack/webpack-cli/blob/master/bin/webpack.js#L46). If we don't exit, yargs will warn saying these flags are unknown and the normal cli flow will follow, with this return, the other file can strip those flags, spawn another webpack-cli process and work as intended.

PlayMa256 replied:
Tests added. Let me know if they aren't enough. I've tried to add as many flags as possible.

ev1stensberg replied:
Okay great, thanks!

ematipico commented:
@ev1stensberg do you have any idea why appveyour goes out of memory during the tests?
